src/agent/initdata-processor/src/initdata.rs

```diff
@@ -0,0 +1,117 @@
+use std::{
+    os::unix::fs::FileTypeExt,
+    path::{Path, PathBuf},
+};
+
+use anyhow::Result;
+use async_compression::tokio::bufread::GzipDecoder;
+use slog::{error, info, Logger};
+use tokio::io::{AsyncReadExt, AsyncSeekExt};
+use tokio::task::JoinHandle;
+
+/// Magic number of initdata device
+pub const INITDATA_MAGIC_NUMBER: &[u8] = b"initdata";
+
+/// It's designed to be run in a separate tokio task to check if a the potential device is the initdata device.
+async fn check_initdata_device(logger: Logger, path: PathBuf) -> Result<Option<String>> {
+    let metadata = match tokio::fs::metadata(&path).await {
+        Ok(m) => m,
+        Err(_) => return Ok(None),
+    };
+
+    if !metadata.file_type().is_block_device() {
+        return Ok(None);
+    }
+
+    info!(logger, "Initdata find a potential device: `{path:?}`");
+
+    let mut file = match tokio::fs::File::open(&path).await {
+        Ok(f) => f,
+        Err(e) => {
+            error!(
+                logger,
+                "Could not open the potential device `{path:?}`: {e}"
+            );
+            return Ok(None);
+        }
+    };
+
+    let mut magic = [0; 8];
+    match file.read_exact(&mut magic).await {
+        Ok(_) if magic == INITDATA_MAGIC_NUMBER => {
+            let device_path = path.to_string_lossy().into_owned();
+            info!(logger, "Found initdata device {device_path}");
+            Ok(Some(device_path))
+        }
+        _ => {
+            // This covers both the case where magic doesn't match and read errors.
+            // We don't need to bubble up read errors as failures for the whole process.
```

Comment:
I would argue about this. I think especially for security sensitive logic there should be as little ambiguity as possible: it either works or fails, graceful degradation isn't helpful in this case, but it'll complicate debugging and misconfigurations might go undetected.